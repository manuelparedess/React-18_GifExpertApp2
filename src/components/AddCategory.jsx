import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({onNewCategory}) => {
    
    const [inputValue, setinputValue] = useState('');
    
    //Para cambiar lo que esta escrito en el input
    const handleInputChange = (e) => {
        setinputValue(e.target.value);
    }

    //Para manejar el submit del formulario
    const handleSubmit = (e) => {
        e.preventDefault();  //Para prevenir que se recargue la pagina al hacer submit
        
        if (inputValue.trim().length >= 2) {
            //setCategories(cats => [inputValue, ...cats]);  
            //Ya no cambia el estado para que solo lo maneje el componente padre

            onNewCategory( inputValue );
            setinputValue('');
        };
    }

    return (
        //Si hay un elemento que contenga a los demas no hace falta poner el fragment
        <form onSubmit={handleSubmit}>
            <input 
            type='text'
            value={inputValue}
            onChange={ handleInputChange }
            />
        </form>
    )
};


AddCategory.propTypes = {
    //onNewCategory tiene que ser una funcion y mandarse como parametro obligatoriamente
    onNewCategory: PropTypes.func.isRequired
}