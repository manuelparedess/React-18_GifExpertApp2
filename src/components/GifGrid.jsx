import React from 'react'
//import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {

    //El custom hook hace el trabajo pesado asi el componente queda 
    //con un codigo mas corto y limpio 
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>

            {loading && <p>Cargando gifs...</p>}

            <div className='card-grid'>

                {
                    images.map((img) =>
                        <GifGridItem
                            key={img.id}
                            {...img} //Para mandar todas las propiedades de img
                        />
                    )
                }

            </div>
        </>
    );
};


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};