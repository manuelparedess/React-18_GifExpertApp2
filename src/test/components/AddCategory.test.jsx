import { render, screen, fireEvent } from "@testing-library/react";
import { AddCategory } from "../../components/AddCategory";


describe('Pruebas en AddCategoy', () => {  

    test('debe de mostrarse correctamente', () => {  
    
        const { container } = render(<AddCategory onNewCategory={ () => {} } />);
    
        expect(container).toMatchSnapshot();
    });

    test('debe de cambiar el valor de la caja de texto', () => {  
        
        render(<AddCategory onNewCategory={ () => {} } />);

        const input = screen.getByRole('textbox');
        const value = 'Hola Mundo';

        //Ya que handleInputChange recibe e y cambia la caja
        //de texto a e.target.value
        fireEvent.input( input, {
            target: {
                value: value,
            }
        });

        expect( input.value ).toBe(value);

    });

    test('debe de llamar onNewCategory y limpiar la caja de texto', () => {  
        
        const onNewCategory = jest.fn();
        
        render(<AddCategory onNewCategory={ onNewCategory } />);

        const value = 'Hola Mundo';

        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        fireEvent.input( input, {
            target: {
                value: value,
            }
        });
        
        fireEvent.submit( form, { preventDefault: () => {} } );

        expect(onNewCategory).toHaveBeenCalledTimes(1);
        expect(onNewCategory).toHaveBeenCalledWith(value);
        expect( input.value ).toBe('');
    });

    test('no debe de aceptar la busqueda', () => {  

        const onNewCategory = jest.fn();
        
        render(<AddCategory onNewCategory={ onNewCategory } />);

        const form = screen.getByRole('form');
        fireEvent.submit( form, { preventDefault: () => {} } );

        expect(onNewCategory).not.toHaveBeenCalled();
    });

})
