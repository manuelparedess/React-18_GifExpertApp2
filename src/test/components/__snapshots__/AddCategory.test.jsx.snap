// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategoy debe de mostrarse correctamente 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      type="text"
      value=""
    />
  </form>
</div>
`;
